lib/inv_cg_quda.cpp

```diff
@@ -224,10 +224,21 @@ namespace quda {
     const int Np = (param.solution_accumulator_pipeline == 0 ? 1 : param.solution_accumulator_pipeline);
     if (Np < 0 || Np > 16) errorQuda("Invalid value %d for solution_accumulator_pipeline\n", Np);
 
-    // Detect whether this is a pure double solve or not; informs the necessity of some stability checks
-    bool is_pure_double = (param.precision == QUDA_DOUBLE_PRECISION && param.precision_sloppy == QUDA_DOUBLE_PRECISION);
+    // Determine whether or not we're doing a heavy quark residual
+    const bool use_heavy_quark_res = (param.residual_type & QUDA_HEAVY_QUARK_RESIDUAL) ? true : false;
+
+    if (use_heavy_quark_res) {
+      hqsolve(x, b);
+      if (param.is_preconditioner) commGlobalReductionPop();
+      return;
+    }
+
+    // This check is pointless in the current version of the code, but it's being proactively added
+    // just in case HQ residual solves are split into a separate file
+    if (use_heavy_quark_res) errorQuda("The \"vanilla\" CG solver does not support HQ residual solves");
 
     // whether to select alternative reliable updates
+    // if we're computing the heavy quark residual, force "traditional" reliable updates
```

Comment:
This comment should be in the heavy quark path?

Reply:
Ah, no, I just forgot to clean that up. I'll take care of that...